master/static/migrations/20240116155819_add-projects-trigger.tx.up.sql

```diff
@@ -0,0 +1,71 @@
+ALTER TABLE projects ADD COLUMN IF NOT EXISTS seq bigint;
+
+-- sequence for tracking event order
+CREATE SEQUENCE IF NOT EXISTS stream_project_seq START 1;
+
+-- trigger function to update sequence number on row modification
+-- this should fire BEFORE so that it can modify NEW directly.
+CREATE OR REPLACE FUNCTION stream_project_seq_modify() RETURNS TRIGGER AS $$
+BEGIN
+    NEW.seq = nextval('stream_project_seq');
+RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS stream_project_trigger_seq ON projects;
+CREATE TRIGGER stream_project_trigger_seq
+    BEFORE INSERT OR UPDATE OF
+    name, description, archived, created_at, notes, workspace_id, user_id, immutable, state
+                     ON projects
+                         FOR EACH ROW EXECUTE PROCEDURE stream_project_seq_modify();
+
+-- helper function to create a project jsonb object for streaming
+CREATE OR REPLACE FUNCTION stream_project_notify(
+    before jsonb, after jsonb
+) RETURNS integer AS $$
+DECLARE
+    output jsonb = NULL;
+BEGIN
+    IF before IS NOT NULL THEN
+        output = jsonb_object_agg('before', before);
+    END IF;
+    IF after IS NOT NULL THEN
+        IF output IS NULL THEN
+            output = jsonb_object_agg('after', after);
+        ELSE
+            output = output || jsonb_object_agg('after', after);
+    END IF;
+END IF;
+    PERFORM pg_notify('stream_project_chan', output::text);
+return 0;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger function to NOTIFY the master of changes, for INSERT/UPDATE/DELETE.
+CREATE OR REPLACE FUNCTION stream_project_change() RETURNS TRIGGER AS $$
+BEGIN
+    IF (TG_OP = 'INSERT') THEN
+        PERFORM stream_project_notify(NULL, to_jsonb(NEW));
+    ELSEIF (TG_OP = 'UPDATE') THEN
+        PERFORM stream_project_notify(to_jsonb(OLD), to_jsonb(NEW));
```

Comment:
I recently learned that postgres notify messages have a [hard-coded 8000 character limit](https://stackoverflow.com/a/41059797).

This means that it isn't safe to send any user-defined fields through the notify mechanism.  How badly does that break our design?

Reply:
I think this means we have two options, find a way to continue to leverage postgres notify/listen while accounting for it's limitations or use a different tool for monitoring DB changes that is more suited for what we're looking to accomplish.

I think we're inevitably going to run into limitations as this current design scales, with the DB being a bottleneck, and then potentially the pub/sub server itself (however you've already accounted for the latter in extensibility section of the design doc), so we'll eventually need to upgrade to something more robust, whether that's homegrown or provided by a third party tool.

Regardless of what approach we use, the impact is limited to how the server consumes events, and should leave the remainder of the framework untouched (including server/client communication); the client shouldn't be concerned about how the server is getting the information, as long as it's communicated in a consistent way. 

Individual streamable types are should be unchanged, as well, I would argue that we could even keep the current implementation of `<X>CollectStartupMsg()`, since we'd only be seeing a change to how live events are received, not how offline events are collected. 

If we go the route of continuing with postgres, we could break large messages up into smaller pieces and the server could reconstructed the total message prior to sending it to the client (or have the client reconstruct it, which might go against keeping the client simple, but would avoid overdoing it on the server side & keeps preparedMessages restricted).

Reply:
If we are to go with a different tool, something like CDC w/Kafka and still funneling events through our publisher framework would be an option, this would remove the load from the DB from handling event notification & allowing our publisher to still handle the relevant filteriong & keep our existing websocket setup.

Something like rabbitmq would be more difficult sense it is more interested in handling the entire process for event creation to client consumption.

Reply:
Just leaving a comment here so that I can be notified properly.
From a slack conversation with Corban it sounds like the code is being re-worked to deal with the character limit.
If the code is actually ready now, or whenever it does become ready, please re-request a review.
Thanks!